oc-chef-pedant/spec/api/stats_spec.rb

```diff
@@ -0,0 +1,137 @@
+require 'pedant/rspec/common'
+require 'mixlib/shellout'
+require 'base64'
+
+describe "/_stats API endpoint", :stats do
+
+  RESPONSE_TYPE_MAP = {
+    "erlang_vm_time_correction" => "UNTYPED",
+    "erlang_vm_thread_pool_size" => "GAUGE",
+    "erlang_vm_threads" => "UNTYPED",
+    "erlang_vm_smp_support" => "UNTYPED",
+    "erlang_vm_schedulers_online" => "GAUGE",
+    "erlang_vm_schedulers" => "GAUGE",
+    "erlang_vm_process_limit" => "GAUGE",
+    "erlang_vm_process_count" => "GAUGE",
+    "erlang_vm_port_limit" => "GAUGE",
+    "erlang_vm_port_count" => "GAUGE",
+    "erlang_vm_logical_processors_online" => "GAUGE",
+    "erlang_vm_logical_processors_available" => "GAUGE",
+    "erlang_vm_logical_processors" => "GAUGE",
+    "erlang_vm_ets_limit" => "GAUGE",
+    "erlang_vm_statistics_wallclock_time_milliseconds" => "COUNTER",
+    "erlang_vm_statistics_runtime_milliseconds" => "COUNTER",
+    "erlang_vm_statistics_run_queues_length_total" => "GAUGE",
+    "erlang_vm_statistics_reductions_total" => "COUNTER",
+    "erlang_vm_statistics_bytes_output_total" => "COUNTER",
+    "erlang_vm_statistics_bytes_received_total" => "COUNTER",
+    "erlang_vm_statistics_garbage_collection_bytes_reclaimed" => "COUNTER",
+    "erlang_vm_statistics_garbage_collection_words_reclaimed" => "COUNTER",
+    "erlang_vm_statistics_garbage_collection_number_of_gcs" => "COUNTER",
+    "erlang_vm_statistics_context_switches" => "COUNTER",
+    "erlang_vm_memory_system_bytes_total" => "GAUGE",
+    "erlang_vm_memory_processes_bytes_total" => "GAUGE",
+    "erlang_vm_ets_tables" => "GAUGE",
+    "erlang_vm_dets_tables" => "GAUGE",
+    "erlang_vm_memory_bytes_total" => "GAUGE",
+    "erlang_vm_memory_atom_bytes_total" => "GAUGE",
+    "erchef_pooler_queued_requestors_max" => "GAUGE",
+    "erchef_pooler_queued_requestors" => "GAUGE",
+    "erchef_pooler_members_max" => "GAUGE",
+    "erchef_pooler_members_free" => "GAUGE",
+    "erchef_pooler_members_in_use" => "GAUGE",
+    "pg_stat_n_conns" => "GAUGE",
+    "pg_stat_n_active_conns" => "GAUGE",
+    "pg_stat_tidx_blks_hit" => "COUNTER",
+    "pg_stat_tidx_blks_read" => "COUNTER",
+    "pg_stat_toast_blks_hit" => "COUNTER",
+    "pg_stat_toast_blks_read" => "COUNTER",
+    "pg_stat_idx_blks_hit" => "COUNTER",
+    "pg_stat_idx_blks_read" => "COUNTER",
+    "pg_stat_heap_blocks_hit" => "COUNTER",
+    "pg_stat_heap_blocks_read" => "COUNTER",
+    "pg_stat_n_dead_tup" => "GAUGE",
+    "pg_stat_n_live_tup" => "GAUGE",
+    "pg_stat_n_tup_del" => "COUNTER",
+    "pg_stat_n_tup_upd" => "COUNTER",
+    "pg_stat_n_tup_ins" => "COUNTER",
+    "pg_stat_tup_fetch" => "COUNTER",
+    "pg_stat_idx_scan" => "COUNTER",
+    "pg_stat_seq_tup_read" => "COUNTER",
+    "pg_stat_seq_scan" => "COUNTER"
+  }
+
+  # Run command once and cache the value.
+  STATS_PASSWORD =
+    begin
+      cmd = Mixlib::ShellOut.new("chef-server-ctl show-secret opscode_erchef stats_password")
```

Comment:
Alternatively to this, we can assume the password is in the environment:

https://github.com/chef/chef-server/blob/master/oc-chef-pedant/lib/pedant.rb#L79-L80

And then do something like this to feed it into the env at runtime:

https://github.com/chef/chef-server/blob/master/omnibus/files/private-chef-ctl-commands/test.rb#L8-L9

Perhaps also putting a guard around the tests to disable them if the secret doesn't exist.

Reply:
You'd also need to add the stats password here: https://github.com/chef/chef-server/blob/master/dev/defaults.yml#L250-L255